### PR TITLE
Issue #2901556 by WidgetsBurritos: Headless Chrome Screenshot Capture Utility

### PR DESCRIPTION
**Pull request by WidgetsBurritos**
Drupal.org Issue: https://www.drupal.org/project/web_page_archive/issues/2901556

This PR adds chrome support to the screenshot capture utility.  Here's a detailed explanation of what's going on:
- Pulls in Google's puppeteer node package to perform screenshots. 
  - Pulling in [spatie/browsershot 3.12.0](https://github.com/spatie/browsershot/releases/tag/3.12.0) as a PHP wrapper around puppeteer.
  - Node/npm are global settings that are used to install and execute puppeteer. 
    - Note: I didn't place this exclusively in the screenshot capture utility because it seems like node could be used across many capture utilities and I'd rather not have to reinvent the wheel in those cases).
  - I also setup `package.json` and `package-lock.json` that are used in combination with `npm` to install puppeteer. 
- You can now select which browser you want to use when configuring a capture utility.
  - For ease of functional testing, I don't presently have it remove browser dropdown options with missing dependencies. If we decide to make that UX improvement, I'd rather handle that in isolation as there's already a ton of stuff going on in this PR. 
  - I modified screenshot capture dependencies as we no longer have a hard mandate on Phantom here.
  - Additionally the capture methods now throw exceptions if they're missing their respective dependency. The exceptions end up in the system log.
- Consolidated the filename generation for all capture utilities to a `getFileName()` method inside `CaptureUtilityBase`.
  - This method prevents duplicate filenames. 
  - This was important as we needed a way to distinguish between the same URL in the phantom/chrome comparison runs.
  - I swapped both the screenshot and the html capture utility to use this method (Note: this only affects new image paths.  Existing images remain the same). 
- I modified the global config form to handle new system settings.
  - I added a form to specify previously mentioned `node`/`npm` settings. 
  - I added a new overridable `buildSystemSettingsForm()` method option to capture utilities.  This allows capture utilities to not only define "default" settings, but also additional system settings.
  - `phantomjs_path` added as an editable system setting.
- Tests were modified to make sure we could properly add `phantomjs` and `chrome` browser settings.
  - Due to external binary dependencies it's a bit difficult to add actual screenshot generation to the functional tests right now. We can revisit this in the future if we decide we need it. 
- Update hooks: 
  - `wpa_screenshot_capture_update_8003()`: Automatically populate `phantomjs_path` with the value from the `PhantomInstaller` package (if available).
  - `wpa_screenshot_capture_update_8004()`: Set's `phantomjs` as the default browser for existing config entities.
  - `web_page_archive_update_8008()`: Initializes node/npm paths.

### CODE REVIEW

File: config/schema/web_page_archive.schema.yml [hunk 1]
```diff
@@ -2,6 +2,18 @@ web_page_archive.settings:
...
+      mapping:
+        node_path:
+          type: string
+          label: 'NodeJS path'
+        phantomjs_path:
+          type: string
+          label: 'PhantomJS path'
+        chrome_path:
+          type: string
+          label: 'Chrome path'
```
Comment (WidgetsBurritos) [inline config/schema/web_page_archive.schema.yml]: just reminding myself to make sure the schema ultimately matches what gets placed in the form as it is inconsistent right now .

File: modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php [hunk 2]
```diff
@@ -35,6 +36,42 @@ public function capture(array $data = []) {
...
+    $file_name .= '.' . rand();
+    $file_name .= ".{$this->configuration['image_type']}";
+    $file_path = $this->storagePath($entity_id, $data['run_uuid']) . '/' . $file_name;
+    $real_file_path = $file_system->realpath($file_path);
+
+    return [
+      'path' => $file_path,
+      'real_path' => $real_file_path,
+    ];
+  }
```
Comment (WidgetsBurritos) [inline modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php]: Reminder for self:  Unit test this

File: modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php [hunk 3]
```diff
@@ -135,13 +213,26 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
...
       '#default_value' => $this->configuration['background_color'],
+      '#states' => [
+        'visible' => [
+          'select[name="data[browser]"]' => ['value' => 'phantomjs'],
+        ],
+      ],
     ];
+    // As of now this only affects phantomjs. Support for this can be added to
+    // chrome but will require a PR/patch against the spatie/browsershot repo.
+    // This can come at a later time.
```
Comment (WidgetsBurritos) [inline modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php]: Based on the comment here spatie/browsershot#130, if browser.js was slightly modified to allow for time-based waiting, we can maintain support in chrome.  

Depending on time, I may try to do open a PR doing this very thing, but may defer until later I get tied up.

File: modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php [hunk 4]
```diff
@@ -47,17 +84,45 @@ public function capture(array $data = []) {
...
+   * Captures using headless chrome.
+   */
+  public function captureHeadlessChrome(array $data = []) {
+    // Retrieve node/npm path.
+    $config = \Drupal::configFactory();
+    $system_settings = $config->get('web_page_archive.settings')->get('system');
+
+    // Run "npm install" inside module.
+    $wpa_dir = \drupal_get_path('module', 'web_page_archive');
+    $install = "cd {$wpa_dir} && {$system_settings['npm_path']} install";
```
Comment (WidgetsBurritos) [inline modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php]: Does this `npm install` need to happen prior to every capture? Or can we move it out to be a little less frequent?  It doesn't really hurt anything as of now, but I'd like to evaluate my options here.

File: modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php (hunk 3 above)
Comment (WidgetsBurritos) [reply modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php]: I ended up opening a PR and it got merged in: spatie/browsershot#139
This got included in v3.12.0 so I'm setting the library explicitly to that version.

**Comment by WidgetsBurritos** [comment]
image


^ This screenshot shows a before/after of using phantom vs chrome.

File: modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php [hunk 5]
```diff
@@ -153,10 +227,25 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildSystemSettingsForm(array &$form, FormStateInterface $form_state) {
+    $config = \Drupal::configFactory()->get('web_page_archive.wpa_screenshot_capture.settings');
+    $form['phantomjs_path'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('NodeJS path'),
```
Comment (WidgetsBurritos) [inline modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php]: PhantomJS*

File: modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php [hunk 6]
```diff
@@ -47,17 +66,56 @@ public function capture(array $data = []) {
...
+  /**
+   * Captures using headless chrome.
+   */
+  public function captureHeadlessChrome(array $data = []) {
+    // Retrieve node/npm path.
+    $config = \Drupal::configFactory();
+    $system_settings = $config->get('web_page_archive.settings')->get('system');
+
+    if (!file_exists($system_settings['npm_path'])) {
+      throw new \Exception(t('npm could not be found at "@path"', ['@path' => $system_settings['npm_path']]));
```
Comment (WidgetsBurritos) [inline modules/wpa_screenshot_capture/src/Plugin/CaptureUtility/ScreenshotCaptureUtility.php]: `$this->t(`

Same below... (and above too)

File: modules/wpa_screenshot_capture/wpa_screenshot_capture.install [hunk 7]
```diff
@@ -47,3 +48,50 @@ function wpa_screenshot_capture_update_8002() {
   $config_storage = \Drupal::service('config.storage');
   $config_storage->write('web_page_archive.wpa_screenshot_capture.settings', $source->read('web_page_archive.wpa_screenshot_capture.settings'));
 }
+
+/**
+ * Sets phantomjs binary if we've previously used the PhantomBinary installer.
+ */
+function wpa_screenshot_capture_update_8003() {
+  // TODO: Add browser to existing configurations.
```
Comment (WidgetsBurritos) [inline modules/wpa_screenshot_capture/wpa_screenshot_capture.install]: Delete comment.

**Review by bighappyface — APPROVED** [review]
+1 for 14ecec29701e957c40ab04b21f9feea3ead8b9f4

**Pull request merged**